Source/IGListGridCollectionViewLayout.m

```diff
@@ -0,0 +1,422 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <IGListKit/IGListAssert.h>
+#import "IGListGridCollectionViewLayout.h"
+
+#pragma mark - IGListGridLayoutLine
+
+@interface _IGListGridLayoutLine : NSObject
+
+/**
+ The scroll direction of the grid.
+ */
+@property (nonatomic, assign) UICollectionViewScrollDirection scrollDirection;
+
+/**
+ The width of the line (in the sense of scroll direction).
+ */
+@property (nonatomic, assign) CGRect frame;
+
+/**
+ The space remains of the line (in the sense of scroll direction).
+ */
+@property (nonatomic, assign) CGFloat tailSpace;
+
+/**
+ The minimum spacing to use between items.
+ */
+@property (nonatomic, assign) CGFloat minimumInteritemSpacing;
+
+/**
+ The section index of the first item in line.
+ */
+@property (nonatomic, assign) NSInteger headIndex;
+
+/**
+ The sizes to of the items in line.
+ */
+@property (nonatomic, copy) NSMutableArray<NSValue *> *itemSizes;
+
+/**
+ Initialization
+ */
+- (id)initWithMinimumInteritemSpacing:(CGFloat)spacing
+                            headIndex:(NSInteger)headIndex
+                                frame:(CGRect)frame
+                      scrollDirection:(UICollectionViewScrollDirection)direction;
+
+/**
+ Adds item to the tail of the line with index path.
+ 
+ @param size The size of the item to be added.
+ 
+ @return A bool indicates if the item can be added to the line.
+ */
+- (BOOL)addItemToTailWithSize:(CGSize)size;
+
+/**
+ Get attributes of the item at index path.
+ 
+ @param indexPath The index path of the item.
+ 
+ @return The attributes for the item in collection view.
+ */
+
+- (UICollectionViewLayoutAttributes *)attributesForItemAtIndexPath:(NSIndexPath *)indexPath;
+
+/**
+ Get attributes of all the item in line.
+ 
+ @return An array of attributes for all the items in line.
+ */
+
+- (NSArray<UICollectionViewLayoutAttributes *> *)attributesForAllItems;
+
+@end
+
+#pragma mark - IGListGridCollectionViewLayout
+
+@interface IGListGridCollectionViewLayout ()
+
+/**
+ The array for line objects in order of line number.
+ */
+@property (nonatomic, copy, nullable) NSMutableArray<_IGListGridLayoutLine *> *lineCache;
+
+/**
+ The line number for each item in order of index path.
+ */
+@property (nonatomic, copy, nullable) NSMutableArray *lineForItem;
+
+/**
+ The width of in collection view content.
+ */
+@property (nonatomic, assign) CGFloat contentWidth;
```

Comment:
This should be `readonly` since the getter is overridden and we're not using ivars.